src/libcore/str/pattern.rs

```diff
@@ -824,27 +843,51 @@ impl TwoWaySearcher {
         // "Algorithm CP2", which is optimized for when the period of the needle
         // is large.
         if &needle[..crit_pos] == &needle[period.. period + crit_pos] {
-            // short period case
+            // short period case -- the period is exact
+            let byteset = needle[..period].iter()
+                                .fold(0, |a, &b| (1 << (b & 0x3f)) | a);
+
+            // compute a separate critical factorization for the reversed needle
+            // x = u' v' where |v'| < period(x).
+            //
+            // This is sped up by the period being known already.
+            // Note that a case like x = "acba" may be factored exactly forwards
+            // (crit_pos = 1, period = 3) while being factored with approximate
+            // period in reverse (crit_pos = 2, period = 2). We use the given
+            // reverse factorization but keep the exact period.
+            let crit_pos_back = needle.len() - cmp::max(
+                TwoWaySearcher::reverse_maximal_suffix(needle, period, false),
+                TwoWaySearcher::reverse_maximal_suffix(needle, period, true));
+
             TwoWaySearcher {
                 crit_pos: crit_pos,
+                crit_pos_back: crit_pos_back,
                 period: period,
                 byteset: byteset,
 
                 position: 0,
                 end: end,
-                memory: 0
+                memory: 0,
+                // memory_back after which we have already matched
+                memory_back: needle.len(),
             }
         } else {
-            // long period case
-            // we have an approximation to the actual period, and don't use memory.
+            // long period case -- we have an approximation to the actual period,
+            // and don't use memorization.
+
+            let byteset = needle.iter()
+                                .fold(0, |a, &b| (1 << (b & 0x3f)) | a);
+
             TwoWaySearcher {
                 crit_pos: crit_pos,
+                crit_pos_back: crit_pos,
```

Comment:
Could there be a comment here explaining that since the period is long, it is longer than both parts of the original critical factorization, and so that critical factorization is usable efficiently for the reverse case?